### DISCUSSION TITLE
No event handler for

### DISCUSSION THREAD

**Discussion by AckslD** [mfussenegger/nvim-dap#601] [Q&A] [answered]
Hi! I'm having some issues with running `pytest` through `dap`. I'm probably doing something wrong in my config or when starting the test so was wondering if anyone knows what might be wrong or have some other tips do debug what is going wrong.

The relevant part of my config is:
```lua
local dap = require('dap')

dap.adapters.python = {
    type = 'executable',
    command = '/usr/bin/python',
    args = {'-m', 'debugpy.adapter'},
}
```

If I simply run a python script with:
```lua
dap.run({
    name = "Launch file",
    type = 'python',
    request = 'launch',
    program = '${file}',
    pythonPath = get_python_path,
    console = 'integratedTerminal',
})
```
it works as expected.

However, when I try to run a test with:
```lua
dap.run({
    name = "Launch file",
    type = 'python',
    request = 'launch',
    module = 'pytest',
    pythonPath = get_python_path,
    console = 'integratedTerminal',
    args = {'test_some_file.py'},
})
```
it does not work and I get the following in the log:
```
[ WARN ] 2022-07-07T14:59:41Z+0200 ] ...e/nvim/site/pack/packer/opt/nvim-dap/lua/dap/session.lua:832 ]	"No event handler for "	{
  body = {
    connect = {
      host = "127.0.0.1",
      port = 40933
    },
    console = "integratedTerminal",
    module = "pytest",
    name = "Subprocess 1454614",
    python = { "/home/axel/venvs/myvenv/bin/python" },
    pythonPath = "/home/axel/venvs/myvenv/bin/python",
    request = "attach",
    subProcessId = 1454614,
    type = "python"
  },
  event = "debugpyAttach",
  seq = 11,
  type = "event"
}
```

I'm not sure why there is no event handler and how to fix that. Any advice would be highly appreciated :)

(btw the `get_python_path` function above is just:
```lua
local get_python_path = function()
    local venv = os.getenv("VIRTUAL_ENV")
    if venv then
        return string.format("%s/bin/python", venv)
    else
        return '/usr/bin/python'
    end
end
```)

**Comment by mfussenegger** [comment, marked as answer]
Your `args` for `pytest` are probably wrong. You need something like `{'-s', 'modulename::className::methodName'}`

I recommend to use [nvim-dap-python](https://github.com/mfussenegger/nvim-dap-python/). It offers a `test_method()` and a `test_class()` method and supports `pytest`.


The `debugpyAttach` event handler warning is only relevant if your application/test uses multiprocessing. That's a custom event handler from debugpy that's not part of the core debug adapter specification.

**Reply by AckslD** [reply]
Thanks a lot for you reply @mfussenegger.

> Your args for pytest are probably wrong. You need something like {'-s', 'modulename::className::methodName'}

Can I see somehow how the process is actually called? Cause normally `pytest test_some_file.py` would work fine. The `-s` flag is only needed to not capture stdout.

> I recommend to use [nvim-dap-python](https://github.com/mfussenegger/nvim-dap-python/). It offers a test_method() and a test_class() method and supports pytest.

Thanks, I tried to look a bit how it's done there but I'd like to try to do it a bit differently and in particular how to pass other arguments to pytest. But if I can't get it working I might have to switch to `nvim-dap-python`.

> The debugpyAttach event handler warning is only relevant if your application/test uses multiprocessing. That's a custom event handler from debugpy that's not part of the core debug adapter specification.

I see, thanks!

**Reply by mfussenegger** [reply]
> Can I see somehow how the process is actually called? Cause normally pytest test_some_file.py would work fine. The -s flag is only needed to not capture stdout.


I think the closest equivalent to what debugpy does with `module = pytest` would be using `python -m pytest` in a shell.

**Reply by AckslD** [reply]
@mfussenegger turns out the issue seems to have been that I passed `args = {"", "${file}"}` since I had done a `vim.split(args_str)` where `args_str` was `" ${file}"`. Doing `vim.split(args_str, {trimempty = true})` does the job :)

Thanks a lot again for the tips and help!

**Reply by AckslD** [reply]
Although not completely sure why that makes it not work, but it's at least working now :)

**Comment by mfussenegger** [comment]
> Thanks, I tried to look a bit how it's done there but I'd like to try to do it a bit differently and in particular how to pass other arguments to pytest

You could customize that. See the part about a custom runner in https://github.com/mfussenegger/nvim-dap-python#usage
